### ISSUE TITLE
Should Repository be parameterised (object, id type)?

### ISSUE THREAD

**Issue by lloydmeta** [debasishg/frdomain#45]
Currently making my way through the book in detail (awesome btw), and I realised that in Chapter 5, `AccountService` is parameterised on `Account`, `Amount`, and `Balance`, which is awesome, but `AccountRepository` on which it depends is not; I was curious if it should be ?

https://github.com/debasishg/frdomain/blob/48664c35b2e583aeacfcacd5d740687afb2ed1de/src/main/scala/frdomain/ch5/domain/service/AccountService.scala#L17-L35

https://github.com/debasishg/frdomain/blob/48664c35b2e583aeacfcacd5d740687afb2ed1de/src/main/scala/frdomain/ch5/domain/repository/AccountRepository.scala#L11-L21

In effect, doesn't this mean that `AccountService` is coupled to the concrete models due to

https://github.com/debasishg/frdomain/blob/48664c35b2e583aeacfcacd5d740687afb2ed1de/src/main/scala/frdomain/ch5/domain/service/AccountService.scala#L18

?

**Comment by debasishg** [comment]
Thanks for your comments .. 

Regarding `AccountRepository`, it's not an implementation, it's a trait for which u can have multiple implementations. And the trait is injected into `AccountService` through a `Kleisli`, which is a `ReaderT`. So the repository is injected into the service from the environment - hence no coupling. And this is done at the end of the world when u implement the Application https://github.com/debasishg/frdomain/blob/master/src/main/scala/frdomain/ch5/domain/app/app.scala#L57) ..

**Comment by lloydmeta** [comment]
Right, but the algebra of the `AccountRepository` is hard-coded to use the concrete `Account` model

https://github.com/debasishg/frdomain/blob/48664c35b2e583aeacfcacd5d740687afb2ed1de/src/main/scala/frdomain/ch5/domain/repository/AccountRepository.scala#L9

That in turn shows up in the algebra of the `AccountService` through the `Kleisli`

https://github.com/debasishg/frdomain/blob/48664c35b2e583aeacfcacd5d740687afb2ed1de/src/main/scala/frdomain/ch5/domain/service/AccountService.scala#L21

In other words, `AccountService#open` for instance is basically:

```scala
trait AccountService[Account /* this is parameterized, which is good */] { 
  def open(
    no: String,
    name: String,
    rate: Option[BigDecimal],
    openingDate: Option[Date],
    accountType: AccountType
  )(repository: { def store(a: models.Account /* here we leak the concrete Account model */): \/[NonEmptyList[String], Account] }): Valid[Account]
}
```

EDIT: My question essentially boils down to: if we're going through the trouble of parameterising `AcocuntService` on `Account` type, wouldn't it make sense to do so for the `AccountRepository` as well? As it is, it seems clear that the only kind of `Account` that could be used with `AccountService` is the concrete `model.Account`, same for `Amount`.

**Comment by debasishg** [comment]
ah .. I get what you say. You are correct! It would certainly be better to parameterize the algebra of repository with the domain entities as type parameters.
